src/node.cc

```diff
@@ -2479,26 +2483,58 @@ struct node_module* get_linked_module(const char* name) {
   return mp;
 }
 
-// DLOpen is process.dlopen(module, filename).
+// DLOpen is process.dlopen(module, filename, flags).
 // Used to load 'module.node' dynamically shared objects.
 //
 // FIXME(bnoordhuis) Not multi-context ready. TBD how to resolve the conflict
 // when two contexts try to load the same shared object. Maybe have a shadow
 // cache that's a plain C list or hash table that's shared across contexts?
 static void DLOpen(const FunctionCallbackInfo<Value>& args) {
   Environment* env = Environment::GetCurrent(args);
+  bool is_dlopen_error = false;
+  const char *dlerrmsg = nullptr;
   uv_lib_t lib;
 
   CHECK_EQ(modpending, nullptr);
 
-  if (args.Length() != 2) {
-    env->ThrowError("process.dlopen takes exactly 2 arguments.");
+  if (args.Length() < 2) {
+    env->ThrowError("process.dlopen needs at least 2 arguments.");
     return;
   }
 
   Local<Object> module = args[0]->ToObject(env->isolate());  // Cast
   node::Utf8Value filename(env->isolate(), args[1]);  // Cast
-  const bool is_dlopen_error = uv_dlopen(*filename, &lib);
+
+#ifdef __POSIX__
+  int flags = 0;
+  if (args.Length() > 2) {
+    if (!args[2]->IsInt32() && !args[2]->IsUndefined()) {
+      return env->ThrowTypeError("dlopen flags must be undefined or integer.");
+    }
+    if (args[2]->IsInt32())
+      flags = args[2]->Int32Value();
```

Comment:
> BTW, Isn't the first `CHECK(args[2]->IsInt32());` redundant since we are using a Maybe type?

Kind of. The `Maybe` is used to indicate errors that happen during a possible conversion to the JS `number` type (for example because calling `[Symbol.toPrimitive]` on `args[2]` throws an exception), not to tell you whether `args[2]` is an Int32 or not.

Basically, there are two ways to do this:
1. If you use `CHECK(args[2]->IsInt32());`, that means you know that `args[2]` is an Int32, and you can use `args[2].As<Int32>()->Value()` to get its `int32_t` interpretation.
2. If you use `args[2]->Int32Value(env->context())`, that means that `args[2]` can be any kind of JS value, which will converted to a `number` based on JavaScript’s conversion rules. Because that operation can run userland code, it can fail, and in that case you get an empty `Maybe<int32_t>`.